### PR TITLE
new lint that detects useless match expression

### LINKED ISSUE #7040: Remove unnecessary match
### What it does

*What does this lint do?*

Remove unnecessary match for error handling when function signatures are the same.

### Categories (optional)

- Kind: `style`, `correctness` ?

*What is the advantage of the recommended code over the original code*

Not sure if there are any performance improvements. The code is more readable and shorter though. This lint also applies for any function pair that returns the same enum, like `Option<T>` for example.

### Drawbacks

None.

### Example

```rust
// For example, a function has the signature:
fn fname(...) -> Result<T, E> { ... }

// and it is called in another function with the same signature:
fn fname2(...) -> Result<T, E> {
    // This is obviously unnecessary
    match fname(...) {
        Ok(s) => Ok(s),
        Err(e) => Err(e),
    }
}
```

Could be written as:

```rust
// For example, a function has the signature:
fn fname(...) -> Result<T, E> { ... }

// and it is called in another function with the same signature:
fn fname2(...) -> Result<T, E> {
    // Much better
    fname(...)
}
```

#### Real world example
```rust
async fn destructure_reqwest_response(res: Response) -> Result<Generic, InternalError> {
    ...
}

async fn get_count(...) -> Result<Generic, InternalError> {
    ...
    // Changed this later. Even clippy::pedantic did not detect this, so I assume that a lint is missing.
    match res {
        Ok(r) => match destructure_reqwest_response(r).await {
            Ok(r) => Ok(r),
            Err(e) => Err(e),
        },
        Err(_) => Err(INTERNALERROR),
    }
    ...
}
```

### PR DESCRIPTION
**Pull request by J-ZhengLi**
fixes #7040 

changelog: Add new  lint [`needless_match`] under complexity lint group

### CODE REVIEW

**Comment by J-ZhengLi** [comment]
Finally got some time to get back to this PR!

I could use some suggestions for my code, such as any missing cases, improvement for readability etc, thanks~ @rustbot ready

File: clippy_lints/src/matches/mod.rs [hunk 1]
```diff
@@ -566,6 +567,49 @@ declare_clippy_lint! {
...
+    /// fn foo() -> Result<(), i32> {
+    ///     result
+    /// }
+    ///
+    /// fn bar() -> Option<i32> {
+    ///     option
+    /// }
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub NOP_MATCH,
```
Comment (llogiq) [inline clippy_lints/src/matches/mod.rs:608]: NOP is uncommon vocabulary. I'd suggest naming it `needless_match` or `match_identity` (as we already have the quite similar `map_identity`).

File: clippy_lints/src/matches/mod.rs [hunk 2]
```diff
@@ -566,6 +567,49 @@ declare_clippy_lint! {
...
+    correctness,
```
Comment (llogiq) [inline clippy_lints/src/matches/mod.rs:609]: I don't see this as a correctness lint. At best, it's suspicious, or complexity.

File: tests/ui/nop_match.rs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+#[derive(Clone, Copy)]
+enum Choice {
+    A,
+    B,
+    C,
+    D,
+}
+
+fn useless_match(x: i32) {
```
Comment (llogiq) [inline tests/ui/nop_match.rs:14]: How about adding a test case with a `x => x` arm?

**Review by llogiq — COMMENTED** [review]
This looks good! :+1: The naming should be more consistent, and we should reconsider the category, and I have a suggestion for further test cases, but otherwise this is merge-worthy.

File: tests/ui/nop_match.rs (hunk 3 above)
Comment (J-ZhengLi) [reply tests/ui/nop_match.rs]: That's a good point, I didn't check for that in my code. :D 

Also, that reminds me, should I be checking `ref x => *x` as well? Or is that too much?

**Comment by llogiq** [comment]
Thank you!

@bors r+

**Comment by J-ZhengLi** [comment]
> Thank you!
> 
> @bors r+

Thank you so much for the approval! But before I change the PR description which possiblly causing this being merged, may I request one last review for the next commit I'm about to push, I have taken your suggestion of adding a `x => x` check along with an extra. 😃

**Pull request merged**